plugins/reporters/spdx/src/main/kotlin/Extensions.kt

```diff
@@ -154,6 +156,26 @@ internal fun Package.toSpdxPackage(
         .filterExcluded()
         .filter(LicenseView.ONLY_DETECTED)
 
+    val declaredPackageLicenses = declaredLicensesProcessed.toSpdxDeclaredLicense()
+    val foundPackageLicenses = declaredPackageLicenses.takeUnless {
+        it == SpdxConstants.NOASSERTION && type == SpdxPackageType.VCS_PACKAGE
+    } ?: run {
+        // If there are of declared licenses and the package type is a VCS, which never have metadata, fall back to
+        // determine the package's license from what is detected in root license files.
+        val patterns = LicenseFilePatterns.getInstance()
+
+        val detectedPackageLicenses = detectedLicenses.filter { licenseInfo ->
+            licenseInfo.locations.any {
+                FileMatcher.match(patterns.allLicenseFilenames, it.location.path, ignoreCase = true)
+            }
+        }
```

Comment:
Why have you decided not to use the root license matcher heuristic as-is?

Reply:
Because after having a look at it again, the root license matcher (whose name BTW is confusing, IMO) does something slightly different: For any given directory / file in the tree, it determines which license files along the path to the root contribute to the licensing of that directory / file. That's more than we need here: It's enough to really only look at the root directory. So I didn't see any code simplification by using the root license matcher.

Reply:
I was aware of the workings of that root license matcher heuristic, I still think it makes sense to use it for the same rationale as why we're using it in ORT.

Reply:
I've reconsidered this, but using the `PathLicenseMatcher` (as it is called now) both adds more code on the caller side, and adds computational overhead that is unneeded for this specific use-case.

Reply:
In what regard does it add more code? Why is the computational overhead problematic?

Would the using `PathLicenseMatcher` yield a different output?

Reply:
> In what regard does it add more code?

The API, i.e. `getApplicableLicenseFindingsForDirectories()`, is not convenient for my use-case. I do not want to operate on raw `LicenseFinding`s, but on `ResolvedLicense`s / `ResolvedLicenseInfo`, to align with the code for `licenseInfoFromFiles` and to easy apply license choices (also see your valid [comment](https://github.com/oss-review-toolkit/ort/pull/9301#discussion_r1824049470) below).

> Why is the computational overhead problematic?

Computational *overhead* (i.e. if you do way more stuff than needed) is always problematic. And since we had performance issues with huge SPDX expressions in the past, I'm simply careful in this rough area.

> Would the using `PathLicenseMatcher` yield a different output?

I don't know (and I don't intend to find out) as I got stuck in the middle of writing a `PathLicenseMatcher`-based implementation, which grew too complex for my taste.

Reply:
> Computational overhead (i.e. if you do way more stuff than needed) is always problematic.

IMO, this does not help moving the discussion forward, because there both, are pros and cons to it.

> The API, i.e. getApplicableLicenseFindingsForDirectories(), is not convenient for my use-case. I do not want to operate on raw LicenseFindings, but on ResolvedLicenses / ResolvedLicenseInfo

If you look at the `ResolvedLicenseInfo` you can see that it actually computes the root license files.
We could simply extend the `RresolvedLicenseInfo` to contain that information. This way you can simply operate on `ResolvedLicenseInfo`, as you preferrred.

What do you think?

Reply:
> If you look at the `ResolvedLicenseInfo` you can see that it actually computes the root license files.

I'm unable to find such code in `ResolvedLicenseInfo` (i.e. there is no use of `LicenseFilePatterns`). Please provide a permalink to what you are referring to.

> There is createLicenseFileInfo(), does this function actually exactly do what you want

I'm uncertainly as I'm having a hard time understanding what that function does. The only thing I want is to get the `ResolvedLicenseInfo` (not `ResolvedLicenseFileInfo`) for all files that are configured for `LicenseFilePatterns`, which is exactly what the above four lines of code do.

Reply:
@fviernau after our discussion I've now changed the code so that is uses the `PathLicenseMatcher` to also work correctly if a package is hosted in the sub-directory of a VCS that has its own license file.